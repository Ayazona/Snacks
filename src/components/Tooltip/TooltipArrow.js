import React, { PureComponent } from 'react'
import PropTypes                from 'prop-types'
import colors                   from '../../styles/colors'

const styles = {
  arrow: {
    position: 'absolute',
    display: 'block',
    height: '12px',
    width: '12px'
  }
}

const RESOLVED_STYLE = {
  primary: {
    backgroundColor: colors.GREEN_500
  },
  secondary: {
    backgroundColor: '#FFF'
  },
  dark: {
    backgroundColor: colors.GRAY_20
  }
}

const RESOLVED_PLACEMENT = {
  top   : {transform: 'rotate(45deg)'},
  bottom: {transform: 'rotate(45deg)'},
  right : {transform: 'rotate(-45deg)'},
  left  : {transform: 'rotate(-45deg)'}
}

const RESOLVE_BORDER_COLOR = {
  primary: colors.GREEN_500,
  secondary: colors.GRAY_74,
  dark: colors.GRAY_20
}

const resolveStylePlacementBorders = (style, placement) => {
  const color = RESOLVE_BORDER_COLOR[style]
  switch (placement) {
    case 'top':
      return {
        borderRight: `1px solid ${color}`,
        borderBottom: `1px solid ${color}`
      }
    case 'bottom':
      return {
        borderLeft: `1px solid ${color}`,
        borderTop: `1px solid ${color}`
      }
    case 'right':
      return {
        borderLeft: `1px solid ${color}`,
        borderTop: `1px solid ${color}`
      }
    case 'left':
      return {
        borderRight: `1px solid ${color}`,
        borderBottom: `1px solid ${color}`
      }
  }
}

class TooltipArrow extends PureComponent {
  static propTypes = {
    position: PropTypes.shape({
      left: PropTypes.number,
      top: PropTypes.number,
      placement: PropTypes.string
    }).isRequired,
    snacksStyle: PropTypes.oneOf(['primary', 'secondary', 'dark'])
  }

  get calculatedStyles() {
    const { position, snacksStyle } = this.props
    const borderStyle = resolveStylePlacementBorders(snacksStyle, position.placement)

    return {
      ...styles.arrow,
      ...RESOLVED_STYLE[snacksStyle],
      ...RESOLVED_PLACEMENT[position.placement],
      ...borderStyle,
      left: position.left,
      top: position.top
    }
  }

  render() {
    return <div style={this.calculatedStyles} />
  }
}

export default TooltipArrow
