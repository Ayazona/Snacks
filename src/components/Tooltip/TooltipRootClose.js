import React, { PureComponent } from 'react'
import { findDOMNode }          from 'react-dom'
import PropTypes                from 'prop-types'

class TooltipRootClose extends PureComponent {
  static propTypes = {
    children: PropTypes.node.isRequired,
    onRootClose: PropTypes.func.isRequired,
    rootCloseEnabled: PropTypes.bool
  }

  static defaultProps = {
    rootCloseEnabled: true
  }

  componentDidMount() {
    const { rootCloseEnabled } = this.props

    if (rootCloseEnabled) {
      this.addEventListeners()
    }
  }

  componentWillReceiveProps(nextProps) {
    const { rootCloseEnabled } = this.props

    if (!rootCloseEnabled && nextProps.rootCloseEnabled) {
      this.addEventListeners()
    } else if (rootCloseEnabled && !nextProps.rootCloseEnabled) {
      this.removeEventListeners()
    }
  }

  componentWillUnmount() {
    const { rootCloseEnabled } = this.props

    if (rootCloseEnabled) {
      this.removeEventListeners()
    }
  }

  addEventListeners() {
    document.addEventListener('click', this.handleMouseClick)
    document.addEventListener('keyup', this.handleKeyUp)
    window.addEventListener('resize', this.handleResize)
  }

  removeEventListeners() {
    document.removeEventListener('click', this.handleMouseClick)
    document.removeEventListener('keyup', this.handleKeyUp)
    window.removeEventListener('resize', this.handleResize)
  }

  handleMouseClick = (e) => {
    const { onRootClose } = this.props
    const target = e.target
    const tooltip = findDOMNode(this)

    if (!tooltip.contains(target)) {
      onRootClose && onRootClose()
    }
  }

  handleKeyUp = (e) => {
    const { onRootClose } = this.props
    if (e.keyCode === 27 && onRootClose) {
      onRootClose && onRootClose()
    }
  }

  handleResize = () => {
    onRootClose && onRootClose()
  }

  render() {
    const { children } = this.props
    return children
  }
}

export default TooltipRootClose
