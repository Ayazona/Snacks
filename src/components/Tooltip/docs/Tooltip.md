Examples:

```js

initialState = {
  size: 'medium',
  placement: 'bottom',
  style: 'primary',
  trigger: 'button',
  isVisible: false,
}

const handlePlacementChange = (placement) => {
  setState({placement: placement})
}

const handleStyleChange = (style) => {
  setState({style: style})
}

const handleSizeChange = (size) => {
  setState({size: size})
}

const handleTriggerChange = (trigger) => {
  setState({trigger: trigger})
}

handleTargetClick = () => {
  setState({isVisible: !state.isVisible})
}

showOnClickContainer = () => {
  return {
    position: 'relative',
    width: '200px',
    height: '200px',
    backgroundColor: 'rgb(20, 110, 4)',
  }
}

const targetDivStyles = {
  position: 'absolute',
  width: '100%',
  height: '100%',
  backgroundColor: 'transparent',
  top: 0,
  left: 0,
}

const targetDiv = () => {
  return <div style={targetDivStyles}/>
}

<div>
  <div style={{display: 'flex', justifyContent: 'space-between'}}>
    <div>
      <h3>Size</h3>
      <RadioGroup name="size" onChange={handleSizeChange}>
        <Radio id="size1" value="small">Small</Radio>
        <Radio id="size2" value="medium">Medium</Radio>
        <Radio id="size3" value="large">Large</Radio>
      </RadioGroup>
    </div>
    <div>
      <h3>Style</h3>
      <RadioGroup name="style" onChange={handleStyleChange}>
        <Radio id="style1" value="primary">Primary</Radio>
        <Radio id="style2" value="secondary">Secondary</Radio>
        <Radio id="style3" value="dark">Dark</Radio>
      </RadioGroup>
    </div>
    <div>
      <h3>Placement</h3>
      <RadioGroup name="placement" onChange={handlePlacementChange}>
        <Radio id="placement1" value="left">Left</Radio>
        <Radio id="placement2" value="right">Right</Radio>
        <Radio id="placement3" value="top">Top</Radio>
        <Radio id="placement4" value="bottom">Bottom</Radio>
      </RadioGroup>
    </div>

    <div>
      <h3>Trigger</h3>
      <RadioGroup name="trigger" onChange={handleTriggerChange}>
        <Radio id="trigger1" value="icon">Icon</Radio>
        <Radio id="trigger2" value="button">Button</Radio>
      </RadioGroup>
    </div>
  </div>

  <div style={{padding: '30px', textAlign: 'center'}}>
    <Tooltip
      target={state.trigger === 'icon' ? <Icon name='info' /> : <Button> Toggle </Button>}
      placement={state.placement}
      size={state.size}
      snacksStyle={state.style}
    >
      {state.size} {state.style} {state.placement}
    </Tooltip>
  </div>
  
  <div style={{display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'space-between'}}>
    <h3>Show Tooltip When Clicking Target</h3>
    <div onClick={handleTargetClick} style={showOnClickContainer()}>
      <Tooltip 
        target={targetDiv()}
        placement={'right'}
        snacksStyle={state.style}
      >
        Tooltip When Clicking Target
      </Tooltip>
    </div>
  </div>
  
  <div style={{display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'space-between'}}>
    <h3>Show Tooltip Onload</h3>
    <div style={showOnClickContainer()}>
      <Tooltip 
        target={targetDiv()}
        placement={'right'}
        snacksStyle={state.style}
        isVisible={true}
      >
        <div style={{width: '100%', height: '100%'}}>
          Tooltip Onload
        </div>
      </Tooltip>
    </div>
  </div>

</div>
```

