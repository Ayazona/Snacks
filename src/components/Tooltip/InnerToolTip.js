import React, { PureComponent } from 'react'
import PropTypes                from 'prop-types'
import Fade                     from '../../components/Transitions/Fade'
import colors                   from '../../styles/colors'
import TooltipArrow             from './TooltipArrow'

const styles = {
  root: {
    position: 'relative'
  },
  arrowPadding: {
    top   : {paddingBottom: '9px'},
    bottom: {paddingTop   : '9px'},
    left  : {paddingRight : '9px'},
    right : {paddingLeft  : '9px'},
  },
  innerContent: {
    textAlign: 'center',
    borderRadius: 4,
    whiteSpace: 'nowrap',
    fontWeight: 600
  },
}

const RESOLVED_COLOR = {
  primary: {
    background: colors.GREEN_500,
    color: '#FFF',
    border: `1px solid ${colors.GREEN_500}`
  },
  secondary: {
    background: '#FFF',
    color: colors.GRAY_46,
    border: `1px solid ${colors.GRAY_74}`
  },
  dark: {
    background: colors.GRAY_20,
    color: '#FFF',
    border: `1px solid ${colors.GRAY_20}`
  }
}

const RESOLVED_SIZE = {
  small: {
    fontSize: '14px',
    padding: '9px 8px'
  },
  medium: {
    fontSize: '16px',
    padding: '9px 16px'
  },
  large: {
    fontSize: '18px',
    padding: '12px 24px'
  }
}

class InnerToolTip extends PureComponent {
  static propTypes = {
    snacksStyle: PropTypes.oneOf(['primary', 'secondary', 'dark']),
    size: PropTypes.oneOf([
      'small',
      'medium',
      'large',
    ]),
    style: PropTypes.shape({
      border: PropTypes.string,
      padding: PropTypes.string,
      boxShadow: PropTypes.srting,
    }),
    arrowStyle: PropTypes.shape({
      border: PropTypes.srting, 
      boxShadowRight: PropTypes.string,
      boxShadowBottom: PropTypes.string,
      boxShadowLeft: PropTypes.string,
      boxShadowTop: PropTypes.string,
    }),
    arrowPosition: PropTypes.shape({})
  }

  static defaultProps = {
    size: 'medium',
    snacksStyle: 'dark'
  }

  get contentStyles() {
    const { size, style, snacksStyle } = this.props
    return {
      ...styles.innerContent,
      ...RESOLVED_SIZE[size],
      ...RESOLVED_COLOR[snacksStyle],
      ...style
    }
  }

  renderArrow() {
    const { arrowPosition, style, arrowStyle, snacksStyle } = this.props
    return (
      <TooltipArrow
        arrowStyle={arrowStyle}
        position={arrowPosition}
        snacksStyle={snacksStyle} />
    )
  }

  render() {
    return (
      <Fade>
        {this.renderArrow()}
        <div style={this.contentStyles}>
          {this.props.children}
        </div>
      </Fade>
    )
  }
}

export default InnerToolTip
