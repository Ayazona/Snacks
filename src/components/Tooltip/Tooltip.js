import React, { PureComponent } from 'react'
import PropTypes                from 'prop-types'
import InnerToolTip             from './InnerToolTip'
import TooltipOverlay           from './TooltipOverlay'

class Tooltip extends PureComponent {
  static propTypes = {
    size: PropTypes.oneOf([
      'small',
      'medium',
      'large',
    ]),
    placement: PropTypes.oneOf([
      'top',
      'left',
      'right',
      'bottom',
    ]),
    target: PropTypes.node.isRequired,
    snacksStyle: PropTypes.oneOf(['primary', 'secondary', 'dark']),
    onDismiss: PropTypes.func,
    onShow: PropTypes.func
  }

  static defaultProps = {
    snacksStyle: 'dark',
    placement: 'bottom',
    size: 'small'
  }

  state = {
    show: false
  }

  handleToggle = () => {
    const {onDismiss, onShow} = this.props
    this.setState({show: !this.state.show}, () => {
      if (this.state.show) {
        onShow && onShow()
      }  else {
        onDismiss && onDismiss()
      }
    })
  }

  handleHideToolTip = () => {
    const {onDismiss} = this.props
    this.setState({ show: false })
    onDismiss && onDismiss()
  }

  renderTriggerElement() {
    const { target } = this.props
    const { show } = this.state

    if (target) {
      return React.cloneElement(target, {
        ref: (node) => {
          this.trigger = node
        },
        onClick: this.handleToggle.bind(this),
        'aria-haspopup': true,
        'aria-expanded': show
      })
    }
  }

  render() {
    const {
      children,
      placement,
      size,
      snacksStyle
    } = this.props

    return (
      <div>
        {this.renderTriggerElement()}
        <TooltipOverlay
          placement={placement}
          target={() => this.trigger}
          show={this.state.show}
          onRootClose={this.handleHideToolTip}
        >
          <InnerToolTip
            size={size}
            snacksStyle={snacksStyle}
          >
            {children}
          </InnerToolTip>
        </TooltipOverlay>
      </div>
    )
  }
}

export default Tooltip
