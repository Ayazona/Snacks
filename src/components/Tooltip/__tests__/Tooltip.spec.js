import React            from 'react'
import ReactTestUtils   from 'react-dom/test-utils'
import renderer         from 'react-test-renderer'
import { StyleRoot }    from 'radium'
import { mount }        from 'enzyme'
import { spy, stub }    from 'sinon'
import Tooltip          from '../Tooltip'

describe('Tooltip', () => {

  it('renders Tooltip properly', () => {
    const tree = renderer.create(
      <StyleRoot>
        <Tooltip
          target={(<div>TRIGGER</div>)}
          placement='right'
          size='small'
          snacksStyle="secondary"
        >
          Right Secondary small
        </Tooltip>
      </StyleRoot>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('expect overlay to appear correctly when shown', () => {
    const tooltip = mount(
      <Tooltip
        target={(<button>TRIGGER</button>)}
        placement='right'
        size='small'
        snacksStyle="secondary"
      >
        Right Secondary small
      </Tooltip>
    )

    const trigger = tooltip.find('button').last()
    trigger.last().simulate('click')

    const tree = renderer.create(tooltip).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should call callbacks', () => {
    const onDismiss = spy()
    const onShow = spy()
    const tooltip = mount(
      <Tooltip
        target={(<button>TRIGGER</button>)}
        placement='right'
        size='small'
        snacksStyle="secondary"
        onShow={onShow}
        onDismiss={onDismiss}
      >
        Callback tooltip
      </Tooltip>
    )

    const trigger = tooltip.find('button').last()
    trigger.last().simulate('click')
    expect(onShow.calledOnce).toBe(true)
    trigger.last().simulate('click')
    expect(onDismiss.calledOnce).toBe(true)
  })
})
