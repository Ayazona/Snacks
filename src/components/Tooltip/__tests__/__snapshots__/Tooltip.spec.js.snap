// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip expect overlay to appear correctly when shown 1`] = `
<div>
  <button
    aria-expanded={true}
    aria-haspopup={true}
    onClick={[Function]}
  >
    TRIGGER
  </button>
  <TooltipOverlay
    onRootClose={[Function]}
    placement="right"
    rootCloseEnabled={true}
    show={true}
    style={
      Object {
        "zIndex": 2345,
      }
    }
    target={[Function]}
  >
    <InnerToolTip
      size="small"
      snacksStyle="secondary"
    >
      Right Secondary small
    </InnerToolTip>
  </TooltipOverlay>
</div>
`;

exports[`Tooltip renders Tooltip properly 1`] = `
<div
  data-radium={true}
>
  <div>
    <div
      aria-expanded={false}
      aria-haspopup={true}
      onClick={[Function]}
    >
      TRIGGER
    </div>
  </div>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;
